lib/Map/Vector/ProtomapsGeojsonSource.ts

```diff
@@ -82,73 +82,133 @@ export class ProtomapsGeojsonSource implements TileSource {
-            transformedGeom = [
-              geom.map((g1) => {
-                g1 = [g1[0] * scale, g1[1] * scale];
-
-                if (bbox.minX > g1[0]) {
-                  bbox.minX = g1[0];
-                }
-
-                if (bbox.maxX < g1[0]) {
-                  bbox.maxX = g1[0];
-                }
-
-                if (bbox.minY > g1[1]) {
-                  bbox.minY = g1[1];
-                }
-
-                if (bbox.maxY < g1[1]) {
-                  bbox.maxY = g1[1];
-                }
-                return new Point(g1[0], g1[1]);
-              })
-            ];
-            numVertices = transformedGeom.length;
-
-            if (f.type === geojsonvt.FeatureTypes.Unknown) return null;
-
-            const geomType = {
-              [geojsonvt.FeatureTypes.Point]: GeomType.Point,
-              [geojsonvt.FeatureTypes.Linestring]: GeomType.Line,
-              [geojsonvt.FeatureTypes.Polygon]: GeomType.Polygon
-            }[f.type];
-
-            const feature: ProtomapsFeature = {
-              props: { ...(f.tags ?? {}) },
-              bbox,
-              geomType,
-              geom: transformedGeom,
-              numVertices
-            };
-
-            return feature;
-          })
-          .filter((f) => f !== null) as ProtomapsFeature[]
+        geojsonVtTileToProtomapsFeatures(tile.features, tileSize)
       );
     }
 
     return result;
   }
 }
+
+export const geomTypeMap = (
+  type: string | null | undefined
+): GeomType | null => {
+  switch (type) {
+    case "Point":
+    case "MultiPoint":
+      return GeomType.Point;
+    case "LineString":
+    case "MultiLineString":
+      return GeomType.Line;
+    case "Polygon":
+    case "MultiPolygon":
+      return GeomType.Polygon;
+    default:
+      return null;
+  }
+};
+
+export function geojsonVtTileToProtomapsFeatures(
+  features: geojsonvt.Feature[],
+  tileSize: number
+): ProtomapsFeature[] {
+  const scale = tileSize / GEOJSON_VT_EXTENT;
+
+  return features
+    .map((f) => {
+      let transformedGeom: Point[][] = [];
+      let numVertices = 0;
+
+      // Calculate bbox
+      const bbox: Bbox = {
+        minX: Infinity,
+        minY: Infinity,
+        maxX: -Infinity,
+        maxY: -Infinity
+      };
+
+      // Multi-polygon
+      if (Array.isArray(f.geometry[0][0])) {
+        // Note: the type is incorrect here
+        const geom = f.geometry as unknown as [number, number][][];
+        transformedGeom = geom.map((g1) =>
+          g1.map((g2) => {
+            const x = g2[0] * scale;
+            const y = g2[1] * scale;
+
+            if (bbox.minX > x) {
+              bbox.minX = x;
+            }
+
+            if (bbox.maxX < x) {
+              bbox.maxX = x;
+            }
+
+            if (bbox.minY > y) {
+              bbox.minY = y;
+            }
+
+            if (bbox.maxY < y) {
+              bbox.maxY = y;
+            }
+            return new Point(x, y);
```

Comment:
Can we extract this to function createPointAndUpdateBbox or two functions, as this is also repeated below